### ISSUE TITLE
Inverse matrix output for zero determinant

### ISSUE THREAD

**Issue by shahsaurabh0605** [SciRuby/nmatrix#436]
invert function gives inverse dense matrix output even if it has zero determinant.

**Comment by GauravTamba** [comment]
Could you share the exact matrix that you're trying to invert? When I try it with a singular matrix it gives me what I think should be the correct error :

```ruby
[14] pry(main)> n = NMatrix.new(3,[1,2,3,4,5,6,2,4,6])
=> 
[
  [1, 2, 3]   [4, 5, 6]   [2, 4, 6] ]
[15] pry(main)> k = n.inverse
ZeroDivisionError: Expected Non-Singular Matrix.
from ~/.rvm/gems/ruby-2.2.2/gems/nmatrix-0.2.0/lib/nmatrix/math.rb:83:in `__inverse__'

```

**Comment by aaditya-thakkar** [comment]
I tried the above matrix it gave me the same ZeroDivisionError but try this matrix:
n= NMatrix.new(3,[1,2,3,4,5,6,7,8,9])
This has determinant zero but still it gives me inverse matrix!

**Comment by shahsaurabh0605** [comment]
I tried the above matrix and it gives me same ZeroDivisionError but try this matrix:
n= NMatrix.new(3,[1,2,3,4,5,6,7,8,9])
It has determinant zero but still it gives an inverse matrix!

**Comment by mohawkjohn** [comment]
Most likely this is because `NMatrix#det` is not equal to `NMatrix#det_exact`. In other words, the approximation being used to compute the determinant is returning a nonzero value, and so it's unaware the function is nonsingular. I'm curious: if you plug in this matrix and try to compute the inverse using `dgetri`, does it give the same answer?

**Comment by GauravTamba** [comment]
I tried out *shahsaurabh0605*'s array input and the inverse function does indeed misbehave however the determinants seem to be correct. 

```ruby
n = NMatrix.new(3, [1,2,3,4,5,6,7,8,9])
=> 
[
  [1, 2, 3]   [4, 5, 6]   [7, 8, 9] ]
pry> n.inverse
=> 
[
  [  643371375338640.9, -1286742750677283.8, 
  643371375338642.2]   [-1286742750677283.8, 
 2573485501354568.5, -1286742750677284.5]  
 [  643371375338642.6, -1286742750677284.5,   643371375338642.2] ]
```

So *inverse* misbehaves with this matrix

```ruby
 pry> n.det
=> 0
 pry> n.det_exact
=> 0
```

**Comment by mohawkjohn** [comment]
That's not what I'm getting:

```
[1] pry(main)> NMatrix.new(3,[1.0,2,3,4,5,6,7,8,9]).det_exact
=> 0.0
[2] pry(main)> NMatrix.new(3,[1.0,2,3,4,5,6,7,8,9]).det
=> 6.661338147750939e-16
```

**Comment by GauravTamba** [comment]
I see, that makes sense. Apologies for missing that.

Is it too expensive in the general case to just compute <NMatrix#det_exact> for the check?

**Comment by shahsaurabh0605** [comment]
I see that the matrix inverse is calculated using Gauss-Jordan elimination method. So is it using NMatrix#det or NMatrix#det_exact to check matrix is non-singular? Because if not, then there must be some other floating point error.

**Comment by GauravTamba** [comment]
This is the line checks for singularity while inverting :

```cpp
        if (matrix[k * (M + 1)] == (DType)(0)) {
          rb_raise(rb_eZeroDivError, "Expected Non-Singular Matrix.");
        }
```
So the determinant does not come into picture here. What I think happens is if the diagonal element on that row cannot be established as a pivot, then a singular matrix is detected. 

Still unsure about the exact cause, but it could be a floating point loss of precision as shahsaurabh0605 already said.

2 simples fixes I can think of right now might be :

a) Check the determinant before computing inverse. (it's as costly as inversion though, `O(n^3)` if I'm correct)

b) Verify that the output matrix multiplied by the input matrix gives us an identity matrix, else raise an exception.

**Comment by GauravTamba** [comment]
I've written a small fix that uses NMatrix#det_exact to reject arrays of size 3 or smaller. It solves the issue for this particular matrix and seems like the best solution (<= 3x3 Matrix Determination calculation isn't very costly) until more singular matrices that return inverses show up.

Would it be okay to send a pull request?

**Comment by mohawkjohn** [comment]
I'm actually not sure about this one. This is a particularly common problem with matrix decompositions. Usually it's solved by adding to the diagonal until the matrix is no longer singular. I'm not convinced it makes sense to address one edge case when it's not really possible to address all of them.

**Comment by GauravTamba** [comment]
I'll try to find more matrices that cause this problem in the meanwhile then. I suggested the above fix because in `det_exact`'s C++ implementation, the computations are hardcoded multiplication for square matrices of shape 3 or less, therefore it would run in about constant time. This particular array returns an inverse in the `lapacke` and `atlas` plugins as well even though lapacke uses LU decomposition to calculate the inverse.

**Comment by mohawkjohn** [comment]
I'd much rather we have an exposed `#inverse_exact` method which relies on `#det_exact`. How about that?

**Comment by shahsaurabh0605** [comment]
Yes, I sent mail indicating the same. Problem is #inverse_exact method is not correctly implemented in the gem. If assigned, I would like to execute it correctly.

**Comment by GauravTamba** [comment]
`#inverse_exact` is implemented in ext/nmatrix/math.cpp already and only works for matrices of size 3 or 2. 

(uppercase comments denoted code blocks edited out)
```cpp
/*
     * Calculate the exact inverse for a dense matrix (A [elements]) of size 2 or 3. Places the result in B_elements.
     */
    template <typename DType>
    void inverse_exact(const int M, const void* A_elements, const int lda, void* B_elements, const int ldb) {
     
       //INITIALIZE
      //SOLVE FOR 1X1
     // SOLVE FOR 2X2
     
      else if (M == 3) {
        // Calculate the exact determinant.
        DType det;
        det_exact<DType>(M, A_elements, lda, reinterpret_cast<void*>(&det));
        if (det == 0) {
          rb_raise(nm_eNotInvertibleError, 
              "matrix must have non-zero determinant to be invertible (not getting this error does not mean matrix is invertible if you're dealing with floating points)");
        }
     
       //CALCULATE INVERSE FOR 3X3 MATRIX
      
      } else if (M == 1) {
        B[0] = 1 / A[0];
      } else {
        rb_raise(rb_eNotImpError, "exact inverse calculation needed for matrices larger than 3x3");
      }
    }

```

I'm not sure if this function is already exposed or not but it doesn't work for matrices larger than 3 since we don't have a fast algorithm for calculating the exact determinant for large matrices yet.

Is the proposal to implement `_exact` for a general square matrix? Sorry for the silly questions and thanks in advance!

**Comment by v0dro** [comment]
@mohawkjohn why even keep this function if the `#inverse` can very well calculate inverses for 3x3 matrices?

**Comment by gau27** [comment]
Here's another singular square matrix that returns an inverse :-

```

**Comment by saru95** [comment]
Any progress on this issue? As far as I can derive from the issue thread above, the problem is caused by matrices of sizes < 3. Can't we just hard-convert the elements to `dtype :float64` for n < 3? Or deprecate `inverse_exact` altogether?

**Comment by mohawkjohn** [comment]
@v0dro We keep it because the computation of exact inverse for 3x3 matrices (e.g., position, velocity) is a common one, and you want it to be exact. `#inverse` is an approximation.

@saru95 This comment gives the fix: https://github.com/SciRuby/nmatrix/issues/436#issuecomment-175048806
